libarchive/archive_read_support_format_tar.c

```diff
@@ -2298,11 +2345,10 @@ pax_attribute(struct archive_read *a, struct tar *tar, struct archive_entry *ent
 					return (err);
 				}
 				else if (key_length == 4 && memcmp(key, "size", 4) == 0) {
-					/* GNU.sparse.size */
+					/* GNU.sparse.size = size of stored entry */
 					if ((err = pax_attribute_read_number(a, value_length, &t)) == ARCHIVE_OK) {
-						tar->realsize = t;
-						archive_entry_set_size(entry, tar->realsize);
-						tar->realsize_override = 1;
+						tar->GNU_sparse_size = t;
```

Comment:
This is the key change:  we now just record what data we see and wait until we've seen all the headers to reconcile it.